### DISCUSSION TITLE
How to make all generated types global?

### DISCUSSION THREAD

**Discussion by christiaanwesterbeek** [kristiandupont/kanel#378] [Q&A]
I want to have all the type be available globally, but I don't know how. I could import and re-export all in some global.d.ts but that's too much work. Any other ideas?

**Comment by kristiandupont** [comment]
Hm, I actually didn't know that it was possible to make types "global". If a type is declared in a file called global.d.ts, what happens?

**Reply by christiaanwesterbeek** [reply]
I am no typescript expert myself, but if you have types in something.d.ts and your tsconfig looks like this, then those types are available in other .ts(x) files without the need to import them.

```
{
    "compilerOptions": {...},
    "include": ["**/*.d.ts"]
}
```

(Just noticed a typeRoots in some tsconfig file. That might be something too...)

**Reply by christiaanwesterbeek** [reply]
To make all generated types available we need to make 3 things happen, I suppose
- Export them as .d.ts instead of .ts
- Stop exporting the types, because that isn't necessary anymore
- Last thing is the trickiest. You can't import anything in the .d.ts files, because as soon as you do that the file becomes a module and defined types stop being global. But there's a workaround to still allow exports. Wrap all the types in a `declare global { ... }`-block, and that problem is fixed.

Some references I found about the 3rd thing:
- https://stackoverflow.com/questions/45420448/how-to-import-external-type-into-global-d-ts-file
- https://stackoverflow.com/questions/71342646/adding-import-statement-to-global-d-ts-destroys-type-and-module-declarations

**Reply by kristiandupont** [reply]
I don't know about typeRoots, but I am pretty sure `include` means which files/folders to include in compilation, not that they are auto-imported?

**Reply by christiaanwesterbeek** [reply]
I have a suspicion that typeRoots could make everything a lot easier.

Hopefully, someone with more typescript expertise chimes in.

**Reply by kristiandupont** [reply]
You could try generating an index file with the index hook (see the example), and then in a `global.d.ts` file simply do `export * from 'models';` (with `models` replaced with whatever folder you used)?
